### PR TITLE
feat(ecosystem): Adds Project Management integration SLOs

### PR DESCRIPTION
**Pull request by GabeVillalobos**
# Adds Project Management/Issues SLO metrics

This PR adds metric coverage to the majority of Project Management features:
- Issue Creation (including alert rules) - src/sentry/api/endpoints/group_integration_details.py
- Outbound status syncing - src/sentry/integrations/tasks/sync_status_outbound.py
- Outbound assignment syncing - src/sentry/integrations/tasks/sync_assignee_outbound.py
- Issue linking - src/sentry/api/endpoints/group_integration_details.py
- Inbound assignment syncing
- Installation metrics for ticketing integrations
- 
## Has Metrics, Needs Testing:
- Inbound status syncing

## Incidental Work:
- Adds new tests for `sync_outbound_assignment` and `sync_outbound_status` tasks, which previously only had endpoint level testing.
- Adds new test for the shared `sync_group_assignee_inbound` utility (src/sentry/integrations/utils/sync.py)

### CODE REVIEW

File: src/sentry/integrations/project_management/metrics.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    LINK_EXTERNAL_ISSUE = "link_external_issue"
+
+    def __str__(self):
+        return self.value.lower()
+
+
+@dataclass
+class ProjectManagementEvent(IntegrationEventLifecycleMetric):
+    action_type: ProjectManagementActionType
+    integration: Integration | RpcIntegration
```
Comment (GabeVillalobos) [inline src/sentry/integrations/project_management/metrics.py:26]: Turns out we don't always have access to integration installations (specifically the inbound assignment code), so I made this use integration/rpc integration.

File: src/sentry/rules/actions/integrations/create_ticket/utils.py [hunk 2]
```diff
@@ -114,27 +118,38 @@ def create_issue(event: GroupEvent, futures: Sequence[RuleFuture]) -> None:
...
+                logger.info(
+                    "%s.rule_trigger.create_ticket.failure",
+                    provider,
+                    extra={
+                        "rule_id": rule_id,
+                        "provider": provider,
+                        "integration_id": integration.id,
+                        "error_message": str(e),
+                        "exception_type": type(e).__name__,
+                    },
```
Comment (GabeVillalobos) [inline src/sentry/rules/actions/integrations/create_ticket/utils.py:144]: It's nice being able to delete these logs in the future

File: tests/sentry/integrations/jira/test_ticket_action.py [hunk 3]
```diff
@@ -1,29 +1,48 @@
...
 
 
+class BrokenJiraMock(MockJira):
+    """
+    Subclass of MockJira client, which implements a broken create_issue method
+    to simulate a misconfigured alert rule for Jira.
+    """
+
+    def create_issue(self, raw_form_data):
+        raise ApiInvalidRequestError("Invalid data entered")
```
Comment (GabeVillalobos) [inline tests/sentry/integrations/jira/test_ticket_action.py:28]: This is gross, but was the easiest way to trigger a failure scenario.

File: src/sentry/api/endpoints/group_integration_details.py [hunk 4]
```diff
@@ -168,56 +172,66 @@ def put(self, request: Request, group, integration_id) -> Response:
...
-        except IntegrityError:
-            return Response({"non_field_errors": ["That issue is already linked"]}, status=400)
+            else:
+                external_issue.update(**defaults)
+
+            installation.store_issue_last_defaults(group.project, request.user, request.data)
+            try:
+                installation.after_link_issue(external_issue, data=request.data)
+            except IntegrationFormError as exc:
+                lifecycle.record_halt(exc)
```
Comment (iamrajjoshi) [inline src/sentry/api/endpoints/group_integration_details.py:217]: are these errors user config errors?

File: src/sentry/integrations/jira/utils/api.py [hunk 5]
```diff
@@ -52,27 +53,34 @@ def handle_assignee_change(
...
 
-    sync_group_assignee_inbound(integration, email, issue_key, assign=True)
+    with ProjectManagementEvent(
+        ProjectManagementActionType.INBOUND_ASSIGNMENT_SYNC, integration=integration
+    ).capture() as lifecycle:
+        assignee_changed = any(
+            item for item in data["changelog"]["items"] if item["field"] == "assignee"
+        )
+        if not assignee_changed:
+            lifecycle.record_halt()
```
Comment (iamrajjoshi) [inline src/sentry/integrations/jira/utils/api.py:64]: could this possibly be a success? are these normal scenarios where jira doesn't send us assignee?

File: src/sentry/integrations/jira/utils/api.py [hunk 6]
```diff
@@ -52,27 +53,34 @@ def handle_assignee_change(
...
+            logger.info("jira.assignee-not-in-changelog", extra=log_context)
+            return
+
+        # If there is no assignee, assume it was unassigned.
+        fields = data["issue"]["fields"]
+        assignee = fields.get("assignee")
+
+        if assignee is None:
+            lifecycle.record_halt()
```
Comment (iamrajjoshi) [inline src/sentry/integrations/jira/utils/api.py:73]: does this mean the ticket was unassigned? why does this have to be a halt?

**Review by iamrajjoshi — COMMENTED** [review]
i just have a few questions, but i think we are kind of aligned on "halt".

i have been thinking of fault as any unexpected condition that isn't user's fault.
with that definition, i have been considering halt as all other non-success conditions, which makes it probably the broadest in definition.

File: src/sentry/api/endpoints/group_integration_details.py (hunk 4 above)
Comment (GabeVillalobos) [reply src/sentry/api/endpoints/group_integration_details.py:217]: Yeah, typically an `IntegrationFormError` is an incorrectly configured Jira ticket creation config.

File: src/sentry/integrations/jira/utils/api.py (hunk 5 above)
Comment (GabeVillalobos) [reply src/sentry/integrations/jira/utils/api.py:64]: That's fair. Tbh though, I'm looking at moving this call into the shared util this calls as there are specific lifecycles we probably want to track there, so this is likely to change.

File: src/sentry/integrations/jira/utils/api.py (hunk 6 above)
Comment (GabeVillalobos) [reply src/sentry/integrations/jira/utils/api.py:73]: Yep, this is actually fine and should be recorded as a success.

File: src/sentry/api/endpoints/group_integration_details.py [hunk 7]
```diff
@@ -168,56 +172,66 @@
...
+            integration=installation.model,
+        ).capture() as lifecycle:
+            try:
+                data = installation.get_issue(external_issue_id, data=request.data)
+            except IntegrationFormError as exc:
+                lifecycle.record_halt(exc)
+                return Response(exc.field_errors, status=400)
+            except IntegrationError as e:
+                lifecycle.record_halt(e)
+                return Response({"non_field_errors": [str(e)]}, status=400)
```
[the remark this answers is not in the document]
Comment (GabeVillalobos) [reply src/sentry/api/endpoints/group_integration_details.py:187]: This isn't new code, this is what allows errors to propagate up to the UI when ticket creation or linking fails on the integrator side (usually a config issue with the ticket contents)

File: src/sentry/integrations/jira/utils/api.py [hunk 8]
```diff
@@ -75,26 +76,40 @@ def handle_assignee_change(
...
+        if not status_changed:
+            logger.info("jira.handle_status_change.unchanged", extra=log_context)
+            return
+
+        try:
+            changelog = next(
+                item for item in data["changelog"]["items"] if item["field"] == "status"
+            )
+        except StopIteration:
+            lifecycle.record_halt("missing-changelog-status", extra=log_context)
```
Comment (iamrajjoshi) [inline src/sentry/integrations/jira/utils/api.py:100]: extreme nit - do we want to put all the reasons in an enum or something?

File: src/sentry/integrations/utils/sync.py [hunk 9]
```diff
@@ -77,48 +81,52 @@ def sync_group_assignee_inbound(
...
+            if user:
+                GroupAssignee.objects.assign(
+                    group,
+                    user,
+                    assignment_source=AssignmentSource.from_integration(integration),
+                )
+                groups_assigned.append(group)
+            else:
+                lifecycle.record_halt("inbound-assignee-not-found", extra=log_context)
+                logger.info("inbound-assignee-not-found", extra=log_context)
```
Comment (iamrajjoshi) [inline src/sentry/integrations/utils/sync.py:128]: we should save a cleanup task for ourselves to remove these redundant loggers

File: src/sentry/rules/actions/integrations/create_ticket/utils.py [hunk 10]
```diff
@@ -114,27 +118,36 @@ def create_issue(event: GroupEvent, futures: Sequence[RuleFuture]) -> None:
...
-                    "provider": provider,
-                },
-            )
 
-            raise
+        with ProjectManagementEvent(
+            action_type=ProjectManagementActionType.CREATE_EXTERNAL_ISSUE,
+            integration=integration,
+        ).capture() as lifecycle:
+            lifecycle.add_extra("provider", provider)
```
Comment (iamrajjoshi) [inline src/sentry/rules/actions/integrations/create_ticket/utils.py:126]: i was planning to add a quick little utility to remove these repeat calls  in my pr here:
https://github.com/getsentry/sentry/pull/80984/files?w=0#r1851251413.

i can update this after

**Review by iamrajjoshi — APPROVED** [review]
lgtm

File: src/sentry/integrations/utils/sync.py (hunk 9 above)
Comment (GabeVillalobos) [reply src/sentry/integrations/utils/sync.py:128]: Good call. Adding it to our SLOs tasks

File: src/sentry/integrations/jira/utils/api.py (hunk 8 above)
Comment (GabeVillalobos) [reply src/sentry/integrations/jira/utils/api.py:100]: Fair! I'll do that now.

File: src/sentry/api/endpoints/group_integration_details.py [hunk 11]
```diff
@@ -168,56 +172,66 @@
...
+
+            installation.store_issue_last_defaults(group.project, request.user, request.data)
+            try:
+                installation.after_link_issue(external_issue, data=request.data)
+            except IntegrationFormError as exc:
+                lifecycle.record_halt(exc)
+                return Response(exc.field_errors, status=400)
+            except IntegrationError as e:
+                lifecycle.record_halt(e)
+                return Response({"non_field_errors": [str(e)]}, status=400)
```
[the remark this answers is not in the document]
Comment (GabeVillalobos) [reply src/sentry/api/endpoints/group_integration_details.py:221]: Same thing here, existing code.

**Review by iamrajjoshi — APPROVED** [review]

**Pull request merged**